activesupport/lib/active_support/hash_with_indifferent_access.rb

```diff
@@ -337,16 +337,32 @@ def transform_values(&block)
       dup.tap { |hash| hash.transform_values!(&block) }
     end
 
-    def transform_keys(*args, &block)
-      return to_enum(:transform_keys) unless block_given?
-      dup.tap { |hash| hash.transform_keys!(*args, &block) }
+    NOT_GIVEN = Object.new # :nodoc:
+
+    def transform_keys(hash = NOT_GIVEN, &block)
+      return to_enum(:transform_keys) if NOT_GIVEN.equal?(hash) && !block_given?
+      dup.tap { |h| h.transform_keys!(hash, &block) }
     end
 
-    def transform_keys!
-      return enum_for(:transform_keys!) { size } unless block_given?
-      keys.each do |key|
-        self[yield(key)] = delete(key)
+    def transform_keys!(hash = NOT_GIVEN, &block)
+      return to_enum(:transform_keys!) if NOT_GIVEN.equal?(hash) && !block_given?
+
+      if hash.nil?
+        super
+      else
+        keys.each do |key|
+          new_key =
+            if NOT_GIVEN.equal?(hash)
+              yield(key)
+            elsif block_given?
+              hash[key] || yield(key)
+            else
+              hash[key] || key
+            end
```

Comment:
You'd probably save a tiny bit of perf on large hashes if you moved these if/elsif outside of the loop.

Reply:
OK. Switched back to that style now. And it looks better indeed. Thank you for the advice!